### PR TITLE
gh-117398: datetime: Support per-interpreter C-API

### PR DESCRIPTION
**Pull request by neonene**
This places C-API structure on a module state, which can be accessed through `PyInterpreterState` or a capsule.

[Doc:](https://docs.python.org/3/c-api/init.html#c.Py_NewInterpreterFromConfig)

> For modules using multi-phase initialization, e.g. PyModule_FromDefAndSpec(),
> a separate module object is created and initialized for each interpreter.

* Issue: gh-117398

### CODE REVIEW

File: Modules/_datetimemodule.c [hunk 1]
```diff
@@ -6924,18 +6935,12 @@ _datetime_exec(PyObject *module)
...
 
     /* At last, set up and add the encapsulated C API */
-    PyDateTime_CAPI *capi = get_datetime_capi();
-    if (capi == NULL) {
-        goto error;
-    }
-    PyObject *capsule = PyCapsule_New(capi, PyDateTime_CAPSULE_NAME,
-                                      datetime_destructor);
+    set_datetime_capi(st);
+    PyObject *capsule = PyCapsule_New(&st->capi, PyDateTime_CAPSULE_NAME, NULL);
```
Comment (Eclips4) [inline Modules/_datetimemodule.c:6939]: Can you explain why do you pass a `NULL` as a capsule destructor?
Is there any reason not to have destructor for this capsule?

File: Modules/_datetimemodule.c [hunk 2]
```diff
@@ -6701,44 +6726,27 @@ static PyMethodDef module_methods[] = {
...
-    capi->Date_FromDate = new_date_ex;
-    capi->DateTime_FromDateAndTime = new_datetime_ex;
-    capi->Time_FromTime = new_time_ex;
-    capi->Delta_FromDelta = new_delta_ex;
-    capi->TimeZone_FromTimeZone = new_timezone;
-    capi->DateTime_FromTimestamp = datetime_fromtimestamp;
-    capi->Date_FromTimestamp = datetime_date_fromtimestamp_capi;
-    capi->DateTime_FromDateAndTimeAndFold = new_datetime_ex2;
-    capi->Time_FromTimeAndFold = new_time_ex2;
+static void
```
Comment (Eclips4) [inline Modules/_datetimemodule.c]: I believe it could be `static inline void`

File: Modules/_datetimemodule.c (hunk 1 above)
Comment (neonene) [reply Modules/_datetimemodule.c:6939]: In this case, the capsule just views a memory allocated and freed by the moduleobject.
https://github.com/python/cpython/blob/bfc57d43d8766120ba0c8f3f6d7b2ac681a81d8a/Objects/moduleobject.c#L739-L740

File: Modules/_datetimemodule.c (hunk 2 above)
Comment (neonene) [reply Modules/_datetimemodule.c]: Applied.

**Comment by neonene** [comment]
No futher experiment here.

**Comment by neonene** [comment]
On second thought, exposing a module state to a user is dangerous.

**Pull request closed without merging**